### PR TITLE
Test SubscriptionOptions::ignore_local_publications

### PR DESCRIPTION
**Pull request by jamoralp**
Signed-off-by: Jose Antonio Moral <joseantoniomoralparras@gmail.com>

Related to https://github.com/ros2/rmw_fastrtps/pull/536

### CODE REVIEW

File: test_rmw_implementation/test/test_subscription.cpp
Comment (hidmic) [inline test_rmw_implementation/test/test_subscription.cpp:535]: @jamoralp consider using `OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT` as in line 750 to ensure publisher destruction on scope exit.

File: test_rmw_implementation/test/test_subscription.cpp
Comment (hidmic) [inline test_rmw_implementation/test/test_subscription.cpp:542]: @jamoralp consider using `OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT` as in line 750 to ensure subscription destruction on scope exit.

File: test_rmw_implementation/test/test_subscription.cpp
Comment (hidmic) [inline test_rmw_implementation/test/test_subscription.cpp:587]: @jamoralp consider using `rmw_wait` instead of `SLEEP_AND_RETRY_UNTIL` (which BTW it is meant to used in cases for which we have no event to wait on yet).

File: test_rmw_implementation/test/test_subscription.cpp
Comment (hidmic) [inline test_rmw_implementation/test/test_subscription.cpp:567]: @jamoralp consider using `rmw_wait` to make sure this part of the test isn't passing simply because the middleware didn't move the message around yet.

File: test_rmw_implementation/test/test_subscription.cpp
Comment (hidmic) [inline test_rmw_implementation/test/test_subscription.cpp:554]: @jamoralp consider using `OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT` as in line 750 to ensure message destruction on scope exit.

File: test_rmw_implementation/test/test_subscription.cpp
Comment (jamoralp) [reply test_rmw_implementation/test/test_subscription.cpp:567]: Ok, sorry for this. I am not very much familiar with the `rmw` and could not think of a better way to achieve this, but I guessed there was some function that I was missing that would help me achieve this. Thank you for your comments

**Review by fujitatomoya — COMMENTED** [review]
lgtm with @hidmic comments.

**Comment by jamoralp** [comment]
Good morning @hidmic @fujitatomoya , this test is failing for both `cyclonedds_rmw` and `fastrtps_rmw`:
* For Cyclone DDS RMW, I guess the problem comes from the fact that they are ignoring publications within their DDS implementation (see [this](https://github.com/ros2/rmw_cyclonedds/blob/17a5a7c2f62531266a08f78bd45b17718f190b86/rmw_cyclonedds_cpp/src/rmw_node.cpp#L1886)). Therefore, the test is throwing a timeout (ret expected: 0 (RMW_RET_OK), actual: 2 (RMW_TIMEOUT)) on `rmw_wait`.
* Talking about Fast DDS RMW, I believe these test have been compiled without the changes introduced in https://github.com/ros2/rmw_fastrtps/pull/536, because for the `ignore_local_publications = true` case, taken should be `false` but, instead, it is `true`. When compiled with the changes I introduced in that pull request and tested in a clean environment in my local machine, they work flawlessly.

Any hints on this? Thank you a lot,
Regards

File: test_rmw_implementation/test/test_subscription.cpp
Comment (hidmic) [inline test_rmw_implementation/test/test_subscription.cpp:593]: @jamoralp consider checking the state of `subscriptions` to make sure both received the message (as described in  [`rmw_wait`](https://github.com/ros2/rmw/blob/a50f3bb1e24ac5338677990d74ab096a9d1776d7/rmw/include/rmw/rmw.h#L2425-L2429) documentation).

Comment (jamoralp) [reply test_rmw_implementation/test/test_subscription.cpp:593]: This will fail for CycloneDDS RMW, since the `ignore_local_publication` flag for their middleware basically blocks the reception of DDS message at a middleware level, not at RMW level; thus, the sub_ignorelocal subscription condition will never be triggered for their RMW. I will check the second subscriber to not bu NULL though.

**Review by hidmic — APPROVED** [review]
LGTM pending green CI

**Comment by jamoralp** [comment]
> LGTM pending green CI

To my understanding, this should happen after a manual trigger, right? Otherwise, there is no way to test it with the changes introduced at https://github.com/ros2/rmw_fastrtps/pull/536

**Comment by hidmic** [comment]
CI up to `test_rmw_implementation`:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14752/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9502/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12416/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14899/)

**Comment by hidmic** [comment]
CI up to  `test_rmw_implementation`:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14779/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9528/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12445/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14926/)

**Comment by hidmic** [comment]
To pass the Rolling PR job (and merge this), we need https://github.com/ros2/rmw_fastrtps/pull/536 released. @MiguelCompany do you want to hold that until https://github.com/ros2/rmw_fastrtps/pull/540 gets in?

**Comment by hidmic** [comment]
Alright, https://github.com/ros/rosdistro/pull/30076 is up. Once it's in, we can re-run the PR job and merge if everything's still fine.

**Comment by hidmic** [comment]
@ros-pull-request-builder retest this please

**Comment by hidmic** [comment]
Alright, all green!

**Pull request merged**
